### DISCUSSION TITLE
Font sizes are broken/inconsistent (how to fix)

### DISCUSSION THREAD

**Discussion by ghost** [Xpra-org/xpra#4400] [Q&A]
So everything was good, every applications loaded up with the correct font/sizes, until yesterday. I've been trying to consolidate some things which meant the machine that is attaching to my XPRA session is back to being multi GPU. After adding a second GPU and another XScreen for it now when I attach to my XPRA session with the Multi GPU/XScreen client some application fonts are normal (as they were) and some are insanely small. i.e. I'm guessing the -more than 1 XScreen- is the issue and requires some extra settings to define things...but what?

I've been searching online and there are only a few old posts about small/wrong fonts (at least that the search engine I used found). Searching on here I find the same few reports that either peter out or don't seem to be applicable. I've also tried various fancy Ai/ML ask/chats with their "Just put butter on it with the -Q flag and pee while spinning in a circle" answers which are useless. [sorry frustration levity] I've read the docs about the DISPLAY bit but from my understanding that is when denoting the XScreen to be simulated on the server, which is now the only machine without multi GPU and only 1 XScreen so from my understanding this doesn't apply. That said I've got some kinda head cold so the sinus brain mush might be my undoing.

Below is an attached example with extra sprinkles. Note in qmmp the media library fonts are all as they should be. Where the playlist and header tab font is way too small. Likewise in qjackctl the fonts are all normal...or _ARE_ they? Notice I've moused over "About" and it has reduced the font size like you see in various UI bits on qmmp and pavucontrol.  There is also inconsistency between launches. Some times some UI elements fonts will be correct, next launch they are tiny. For example if I close and reopen qmmp on the existing session the next launch those *correct* library fonts will be tiny like everything else.

![xpra-example](https://github.com/user-attachments/assets/08ddf12e-897f-4f4d-9280-37293b430b45)

What is/might be causing this odd font inconsistency now and how might it be remedied? What am I missing? (Nothing of relevance shows in the logs either.)

Thanks to whomever takes a peek at this.

*** P.S. I tried something silly which was to try the scaling options from the tray menu. The fonts that render correctly scaled up along with the basic UI and window. The problematic fonts stayed the same size with the added feature of blur. Not sure if this might give a hint to someone that knows better about what's going on but it was interesting.

Since you have locked the discussion...
If you mean "what operating system is used on both the client and server, including the full version details" you may notice if you are going to get pedantic that is a BUG REPORTING guideline...this is NOT a bug report, this is a Discussion.

**Comment by totaam** [comment]
Have you read https://github.com/Xpra-org/xpra/blob/master/docs/Features/DPI.md ?
Please see https://github.com/Xpra-org/xpra/wiki/Reporting-Bugs
A number of issues are simply fixed by using Xdummy - which is very problematic on Debian distros.

**Reply by ghost** [reply]
I'm not on debian and no I didn't read the DPI bit because this doesn't seem to be a DPI issue. I'd have thought if it was DPI all the fonts would be right or wrong, not selectively different. That said icons and most other things are as they should be but if there was DPI scaling gone awry everything should be whack should it not?

**Reply by totaam** [reply]
> if it was DPI all the fonts would be right or wrong, not selectively different.

As per the DPI documentation, there are many sources for the DPI value.

**Reply by ghost** [reply]
OK but again everything has been fine up until the new GPU and added XScreen so why now would the DPI be wrong?

**Reply by ghost** [reply]
I'm not sure if I'm looking at the correct lines but I ran both client and server with the -d dpi and
Server 5760x1080 50Hz (1016x190 mm, dpi=144)
Client :0.0 (1540x292 mm - DPI: 95x94) workarea: 5760x1079 at    0x1

Then for screens
Server
2024-10-23 21:25:47,173  1920x1080 60Hz (509x286 mm, dpi=96) at 1920,0
2024-10-23 21:25:47,173  1920x1080 60Hz (509x286 mm, dpi=96)
2024-10-23 21:25:47,175  1920x1080 60Hz (509x286 mm, dpi=96) at 3840,0

Client
2024-10-23 21:25:46,928     ACR DP-4         1920x1080 at    0x0    (509x286 mm - DPI: 96x96) workarea: 1920x1080
2024-10-23 21:25:46,928     ACR HDMI-0       1920x1080 at 1920x0    (509x286 mm - DPI: 96x96) workarea: 1920x1079 at 1920x1
2024-10-23 21:25:46,928     ACR DP-1         1920x1080 at 3840x0    (509x286 mm - DPI: 96x96) workarea: 1920x1080 at 3840x0

**Reply by totaam** [reply]
Please see "reporting bugs" and clarify what you mean by `XScreen`

**Reply by ghost** [reply]
XScreen is an XScreen as defined in you xorg,conf DISPLAY=:0.0 XScreen 0 DISPLAY=:0.1 XScreen 1 Well unless you ran gnome back when it still enumerated XScreens, dif syntax.

**Reply by totaam** [reply]
xpra is not screen aware, so I'm not sure how this even works!

**Reply by ghost** [reply]
Well I'm trying to read and test and I tried starting things after the client had connected but I still get tiny fonts. XPRA shouldn't need to care about the XScreens because atm the session on the server is on 0.0 (the only XScreen since it's a single GPU machine) and the client attaches on 0.0. I'm just saying the font issue only appeared after adding the second GPU and XScreen for that GPU's monitors. It could be a coincidence and nothing more though. It was simply the only thing I could say changed when the issue popped up as a source of potential issues.

**Reply by ghost** [reply]
I've been fiddling and looks like this has nothing to do with the second XScreen / GPU on the client. If I start something on the host then attach with the host...I get tiny fonts! (The plot thickens!) I also tried using the GUI launcher and disabled screen scaling and set DPI to 96...tiny fonts.

**Comment by ghost** [comment]
I've continued to play with various XPRA settings to do with scaling, dpi, x settings and whatever I can find but if I launch something like l3afpad via XPRA on the host then attach via the host the fonts are not only wrong but always different. Some times the menu fonts are super tiny, some times normal. Some times the editor font is small, normal, large or even ridiculously massive.

Here is a snipped screenshot of 2 instances of l3afpad running over XPRA from the host to the host. It shows the tiny fonts and the inconsistent big fonts.

![l3afpad-snip2](https://github.com/user-attachments/assets/a4e3ce19-1017-45fb-82b2-939063946380)

**Reply by totaam** [reply]
Sounds like you may need to look into `--start-after-connect` so that the virtual display has been configured to match the client's before the application is started.

**Reply by ghost** [reply]
I had read about that and tried it yesterday, it doesn't change anything.

**Reply by totaam** [reply]
It certainly does change things for some applications, but this may not have as much of an effect on Debian - without Xdummy.

**Reply by ghost** [reply]
Again I am _not_ on a Debian distro.

**Reply by totaam** [reply]
> Again I am NOT on a Debian distro.

Again?
https://github.com/Xpra-org/xpra/wiki/Reporting-Bugs

**Reply by ghost** [reply]
So are you saying this is a bug rather than a config issue? There is a reason I started a discussion rather than a bug report but I'll report with the GUI tool then.

**Reply by totaam** [reply]
No, I am saying that I pointed you to that wiki page to save us all some time.
I have no idea what distro you're on or how you installed xpra, etc..

**Reply by ghost** [reply]
I'm on Arch (you didn't ask you just kept saying I was on Debian ;p) xpra is in "extra." 

I just tried the bug reporting tool from the tray as the wiki says and frankly I'm even more confused now. I started another instance of l3afpad and not only is the font even larger for input but if I open the text report it's compiled to send in the bug report it's all wrong.

```
xpra start :100 ssh:user@host/ --exit-with-children --start-child=/usr/bin/vlc
xpra start :100 --exit-with-children --start-child=/usr/bin/vlc

Run from Server
xpra start ssh:user@host --exit-with-children --start-child=/usr/bin/vlc

Start locally
xpra start :10
DISPLAY=:10 "command"

xpra attach ssh:user@host:10
```

![xpra-snip3](https://github.com/user-attachments/assets/28d10535-1b66-49b7-aa52-fe8ab709c97a)

I'm just giving up on this for now, I have more important Ghosts to hunt. Perhaps I'll revisit this when things are less chaotic.

Thanks for your time.

**Reply by totaam** [reply]
```shell
xpra start :10
DISPLAY=:10 "command"
```
As per the documentation, never do that.
Use `xpra start --start=command` so that the command inherits the correct environment.
The arch documentation has been wrong in this subject for the best part of a decade and should be nuked.

**Reply by totaam** [reply]
> `xpra attach ssh:user@host:10`

Again, this syntax is deprecated and should give you warnings.

**Reply by totaam** [reply]
And the information from https://github.com/Xpra-org/xpra/wiki/Reporting-Bugs is still missing

**Reply by ghost** [reply]
I didn't do that, (I said it was wrong) that is what the bug reporting tool gave me. When I launch normally I use
```
xpra start :100 --start-after-connect --start='qmmp' --start '/usr/bin/qjackctl' --start 'env GTK_THEME=Graphite-yellow-Dark-compact pavucontrol -t 1' --start '/usr/bin/geany -t -c $HOME/.config/geany.top' --speaker=no --microphone=no
```

or in this case

```
xpra start :100 --start l3afpad
```

For some reason the bug reporting has that :10 command bit and something about VLC neither of which I did.

**Reply by ghost** [reply]
I'm not sure what info you want from reporting bugs as this is a discussion and I've not gone through things completely to report a bug. More so I was in the process of going through the bug wiki and saw the bug reporting tool as it instructs gave me incorrect info (the :10 and vlc bit I noted).

More so I said I'd leave this here for now.

**Reply by totaam** [reply]
> I didn't do that, (I said it was wrong)

That statement is so unclear, I have no idea what to do with it.

> For some reason the bug reporting has that :10 command bit and something about VLC neither of which I did.

Again, no idea what this refers to.

> I'm not sure what info you want from reporting bugs

Enough to have an idea of where to go.
I'm going to guess that you're not running an up to date version at this point.

> the bug reporting tool as it instructs gave me incorrect info

The tool collects information from the system.
It cannot invent `:10` or `vlc`.

**Reply by ghost** [reply]
I didn't do "any" of what I posted from the bug reporting tool. When I was reading the wiki I tried the bug reporting tool, below is the info it copied to clipboard and saved...none of which is what I did/ran/used.

```
xpra start :100 ssh:user@host/ --exit-with-children --start-child=/usr/bin/vlc
xpra start :100 --exit-with-children --start-child=/usr/bin/vlc

Run from Server
xpra start ssh:user@host --exit-with-children --start-child=/usr/bin/vlc

Start locally
xpra start :10
DISPLAY=:10 "command"

xpra attach ssh:user@host:10
```

"The tool collects information from the system.
It cannot invent :10 or vlc."

Well it did. I've never run vlc over xpra, I don't use :10 and I don't use the syntax xpra attach ssh:user@host:10 (my attach is xpra attach ssh://user@host/100) the above commands are all from the xpra bug reporting tool and does not have anything "real" that I've ran or used for commands.

As for XPRA not being up to date it's Version: xpra v6.2.0-r0

**Reply by totaam** [reply]
None of these strings exist in the xpra source tree.

**Reply by ghost** [reply]
Exist or not that is what the tool gave me, which was again confusing since it's not what I had used or ran. That's the only reason I noted it, because it doesn't reflect anything sensible or that has actually been done. Either way this seems like a non-starter so I'm for the third time saying I'm leaving it.

**Comment by ghost** [comment]
For anyone else finding this in the future...

My setup has changed a lot, same amount of screens just different GPU's running on different machines. However now with single XScreen to single XScreen font sizes are still wrong. --DPI=144 makes it almost normal despite the host and client machines both being 96. XPRA remains borked and discombobulated. Just use X forwarding where applicable.

**Reply by totaam** [reply]
After almost a year, we still don't know what distro you're using. Amongst many other quite statements rather unclear or unrelated to xpra.
My guess is that you are using Debian or Ubuntu - which can't simulate DPI properly.

**Reply by ghost** [reply]
I run Arch, yet you never asked, you just repeatedly assumed Debian. I repeatedly told you I AM NOT on a Debian based distro but you kept insisting I should "not use debian."

**Comment by totaam** [comment]
> yet you never asked,

False, my comment above from October 2024 points to:
> Please see https://github.com/Xpra-org/xpra/wiki/Reporting-Bugs
